accepted/2023/feature-switch-attributes.md

```diff
@@ -0,0 +1,719 @@
+
+- Terminology: "feature" vs "capability"
+
+  We use "feature" instead of "capability" because "capability" suggests a quality of the underlying platform or runtime, while "feature" seems slightly more general, and can refer to settings toggled in the application config. This lines up with what we have been describing as "feature switches", and with the naming of `RuntimeFeature.IsDynamicCodeSupported` and similar APIs.
+
+- Terminology: "check" vs "switch"
+
+  We use `FeatureSwitch` in essentially the same way that the capability API draft uses `CapabilityCheck`. Maybe `FeatureCheck` would be a better name for the attribute, since the "switch" is really the MSBuild property that can be toggled, while the attribute should indicate that the property is a check for that setting.
+
+- Representing `RequiresDynamicCodeAttribute` as a feature/capability
+
+  The capability API draft mentions the possibility of using a capability model for `RequiresDynamicCodeAttribute`, just as we suggest here. This is the same idea.
+
+- Tie-in to feature switches
+
+  The capability API draft doesn't include a tie-in to the feature names.
+
+- API shape: target of `CapabilityCheck` vs `FeatureSwitch`
+
+  The capability API draft places `CapabilityCheck` on the capability attribute definition, for example:
+
+  ```csharp
+  [CapabilityCheck(typeof(RuntimeFeature), nameof(RuntimeFeature.IsDynamicCodeSupported))]
+  public sealed class RequiresDynamicCodeAttribute : CapabilityAttribute
+  {
+      public RequiresDynamicCodeAttribute();
+  }
+  ```
+
+  We instead suggest placing the equivalent `FeatureSwitch` attribute on the property definition:
+
+  ```csharp
+  public sealed class RuntimeFeature
+  {
+      [FeatureSwitch(typeof(RequiresDynamicCodeAttribute))]
+      public bool IsDynamicCodeSupported => // ...
+  }
+  ```
+
+  Both models allow multiple "feature switch" properties to be defined by the same feature, but the latter can prevent the same property from being marked as the "feature check" for multiple different features that could potentially conflict, if we set `AllowMultiple = false` on `FeatureSwitchAttribute`.
+
+- API shape: target of `CapabilityGuard` vs `FeatureGuard`
+
+  These are the same idea, and are represented very similarly in both models, targeting the guard property, with `AllowMultiple = true` to support a property that guards multiple features. The capability API draft lists the attribute targets as `AttributeTargets.Method | AttributeTargets.Property | AttributeTargets.Field`. We might want to start with just `AttributeTargets.Property`.
+
+## Comparison with platform compatibility analyzer
+
+The platform compatibility analyzer is semantically very similar to the behavior described here, except that it doesn't come with ILLink/ILCompiler support for removing removing branches that are unreachable when publishing for a given platform.
```

Comment:
> except that it doesn't come with ILLink/ILCompiler support for removing removing branches that are unreachable when publishing for a given platform

This is kind of implicitly working because `OperatingSystem.Is<OS>()` APIs are hardcoded to return false for the non-matching platforms so ILLink/ILCompiler constant propagation takes care of that.